src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,146 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
-  auto start = absl::Now();
-  auto reconnected = false;
-  while (absl::Now() - start < timeout_s) {
-    if (disconnected_) {
+  if (get_server_address_func_(&address)) {
+    // After GCS is restarted, the gcs client will reestablish the connection. At
+    // present, every failed RPC request will trigger `ReconnectGcsServerAsync`. In order
+    // to avoid repeated connections in a short period of time, we add a protection
+    // mechanism: if the address does not change (meaning gcs server doesn't restart),
+    // the connection can be made at most once in
+    // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
+    if (last_reconnect_address_ == address &&
+        (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
+            RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
+      RAY_LOG(DEBUG)
+          << "Repeated reconnection in "
+          << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
+          << " milliseconds, return directly.";
+      ExecuteReconnectedCallbacks();
+      reconnect_in_progress_ = false;
       return;
     }
-    if (get_server_address_func_(&address)) {
-      // After GCS is restarted, the gcs client will reestablish the connection. At
-      // present, every failed RPC request will trigger `ReconnectGcsServer`. In order to
-      // avoid repeated connections in a short period of time, we add a protection
-      // mechanism: if the address does not change (meaning gcs server doesn't restart),
-      // the connection can be made at most once in
-      // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
-      if (last_reconnect_address_ == address &&
-          (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
-              RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
-        RAY_LOG(DEBUG)
-            << "Repeated reconnection in "
-            << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
-            << " milliseconds, return directly.";
-        return;
-      }
-
-      RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
-                     << address.second;
-      if (Ping(address.first, address.second, 100)) {
-        // If `last_reconnect_address_` port is -1, it means that this is the first
-        // connection and no log will be printed.
-        if (last_reconnect_address_.second != -1) {
-          RAY_LOG(INFO) << "Reconnected to GCS server: " << address.first << ":"
-                        << address.second;
-        }
-        reconnected = true;
-        break;
-      }
-    }
-    std::this_thread::sleep_for(
-        std::chrono::milliseconds(kGCSReconnectionRetryIntervalMs));
+
+    RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
+                   << address.second;
+    gcs_rpc_client_->ResetHeartBeatInfoClient(address.first, address.second,
```

Comment:
Why do we need it here? Can we just make it a part of `gcs_rpc_client_->Reset(address.first, address.second, *client_call_manager_);`?

Reply:
The main intention is to update the ip:port of `heartbeat_info_grpc_client_` independently since the ip:port of gcs_server may have changed. `gcs_rpc_client_->Reset` reset all grpc clients including the `heartbeat_info_grpc_client_`, actually i am not sure whether there is any additional impact.